com.oracle.truffle.r.runtime/src/com/oracle/truffle/r/runtime/data/RAltIntVectorData.java

```diff
@@ -0,0 +1,286 @@
+import com.oracle.truffle.r.runtime.nodes.altrep.AltrepSetElementAtNode;
+import com.oracle.truffle.r.runtime.ops.na.NACheck;
+
+@ExportLibrary(VectorDataLibrary.class)
+public class RAltIntVectorData implements TruffleObject, VectorDataWithOwner {
+    protected final RAltRepData altrepData;
+    protected final AltIntegerClassDescriptor descriptor;
+    private boolean dataptrCalled;
+    private RIntVector owner;
+
+    public RAltIntVectorData(AltIntegerClassDescriptor descriptor, RAltRepData data) {
+        this.altrepData = data;
+        this.descriptor = descriptor;
+        this.dataptrCalled = false;
+        assert hasDescriptorRegisteredNecessaryMethods(descriptor):
+                "Descriptor " + descriptor.toString() + " does not have registered all necessary methods";
+    }
+
+    private boolean hasDescriptorRegisteredNecessaryMethods(AltIntegerClassDescriptor descriptor) {
+        return descriptor.isLengthMethodRegistered() && descriptor.isDataptrMethodRegistered();
+        /* TODO: && descriptor.isUnserializeMethodRegistered(); */
+    }
+
+    public AltIntegerClassDescriptor getDescriptor() {
+        return descriptor;
+    }
+
+    public RAltRepData getAltrepData() {
+        return altrepData;
+    }
+
+    public Object getData1() {
+        return altrepData.getData1();
+    }
+
+    public Object getData2() {
+        return altrepData.getData2();
+    }
+
+    public void setData1(Object data1) {
+        altrepData.setData1(data1);
+    }
+
+    public void setData2(Object data2) {
+        altrepData.setData2(data2);
+    }
+
+    @Override
+    public void setOwner(RAbstractVector newOwner) {
+        owner = (RIntVector) newOwner;
+    }
+
+    private RIntVector getOwner() {
+        assert owner != null;
+        return owner;
+    }
+
+    public void setDataptrCalled() {
+        dataptrCalled = true;
+    }
+
+    public boolean wasDataptrCalled() {
+        return dataptrCalled;
+    }
+
+    @ExportMessage
+    public NACheck getNACheck(@Cached() NACheck na) {
+        na.enable(false);
+        return na;
+    }
+
+    @SuppressWarnings("static-method")
+    @ExportMessage
+    public final RType getType() {
+        return RType.Integer;
+    }
+
+    @ExportMessage(limit = "1")
+    public int getLength(@Cached("create()") AltrepLengthNode lengthNode) {
+        return (int) lengthNode.execute(getOwner());
+    }
+
+    @ExportMessage(limit = "1")
+    public RIntArrayVectorData materialize(@Cached("create()") AltrepRFFI.AltIntDataptrNode dataptrNode,
+                                           @Cached AltrepLengthNode lengthNode) {
+        int length = (int) lengthNode.execute(getOwner());
+        long dataptrAddr = dataptrNode.execute(getOwner(), true);
+        int[] newData = new int[length];
+        NativeMemory.copyMemory(dataptrAddr, newData, ElementType.INT, length);
+        // TODO: complete=true?
+        return new RIntArrayVectorData(newData, true);
```

Comment:
If we do not know anything about the data, it complete should be `false`. You can use more idiomatic `RDataFactory.INCOMPLETE_VECTOR` (but the existing code is not consistent in using that unfortunately)